### PR TITLE
Add option to publish via a different URI

### LINKED ISSUE #45: Allow anonymous access or base url in index to be overridden
Helm-s3 is really useful for uploading and managing helm charts via in S3. 
However, I want my repo to be public, so ideally the provider chain would fall back to anonymous access to s3, or a baseurl could be specified to be used in the index.yaml instead?

### PR DESCRIPTION
**Pull request by steven-sheehy**
Fixes #45 

This PR implements the proposal in PR #56, but addresses the concerns that I had with the proposal. It does so by maintaining the remote and local indexes separately such that the remote one has the published URI (HTTP) while the local always has the S3 URIs. Besides solving the problem with authenticated repositories while still supporting the other use cases, I think it is more intuitive that the Helm S3 plugin continues to use S3 as its protocol regardless of how the repository user consumes the publish repository.

I also merged the publish command into a `helm s3 reindex --publish` since the publish command was effectively a reindex command with a different publish URI.

### CODE REVIEW

File: cmd/helms3/delete.go [hunk 1]
```diff
@@ -56,16 +59,35 @@ func (act deleteAction) Run(ctx context.Context) error {
...
 		return errors.WithMessage(err, "upload new index to s3")
 	}
 
-	if err := idx.WriteFile(repoEntry.Cache, 0644); err != nil {
+	localIndexFile, err := repo.LoadIndexFile(repoEntry.Cache)
+	if err != nil {
+		return err
+	}
+
+	localIndex := &index.Index{localIndexFile}
```
Comment (golangcibot) [inline cmd/helms3/delete.go]: `index.Index` composite literal uses unkeyed fields (from `govet`)

File: cmd/helms3/push.go [hunk 2]
```diff
@@ -158,11 +172,17 @@ func (act pushAction) Run(ctx context.Context) error {
 	}
 
 	if !act.dryRun {
-		if err := storage.PutIndex(ctx, repoEntry.URL, act.acl, idxReader); err != nil {
+		if err := storage.PutIndex(ctx, repoEntry.URL, publishURI, act.acl, idxReader); err != nil {
 			return errors.WithMessage(err, "upload index to s3")
 		}
 
-		if err := idx.WriteFile(repoEntry.Cache, 0644); err != nil {
+		localIndex := &index.Index{cachedIndex}
```
Comment (golangcibot) [inline cmd/helms3/push.go]: `index.Index` composite literal uses unkeyed fields (from `govet`)

File: cmd/helms3/delete.go [hunk 3]
```diff
@@ -29,7 +31,8 @@ func (act deleteAction) Run(ctx context.Context) error {
 	storage := awss3.New(sess)
 
 	// Fetch current index.
-	b, err := storage.FetchRaw(ctx, repoEntry.URL+"/index.yaml")
+	indexUri := repoEntry.URL + "/index.yaml"
```
Comment (golangcibot) [inline cmd/helms3/delete.go]: var indexUri should be indexURI (from `golint`)

**Comment by steven-sheehy** [comment]
I've added integration tests for published repositories. I also refactored the integration tests to remove the unnecessary `if [ $? -ne 0 ]` checks since `-e` will already catch non-zero return codes and to print out the failed test case.

**Comment by steven-sheehy** [comment]
@hypnoglow Anything I can do to move this PR along?

**Comment by jgangemi** [comment]
any possible movement on this? i could also really use this functionality.

**Comment by steven-sheehy** [comment]
@jgangemi Until this merged, you're welcome to use my forked build:

`helm plugin install https://github.com/steven-sheehy/helm-s3.git --version v0.9.0`

**Comment by jgangemi** [comment]
@steven-sheehy awesome, thanks! 

what's the behavior if i still want to access via `s3`? will both protocols work to read the charts or am i locked into one or the other.

**Comment by steven-sheehy** [comment]
@jgangemi Nope, you can use both concurrently, even from the same client. Just add the s3 and http as separate repos via `helm repo add`. The s3 repo can be used for read/write while the http can do reads.

**Comment by jgangemi** [comment]
@steven-sheehy awesome, thanks again!!

**Comment by jgangemi** [comment]
hey @steven-sheehy, you didn't touch anything w/ the `push` that would have affected `*` being resolved, eg:

`helm s3 push --dry-run target/* repo`

this used to work in an older version and doesn't ask of `0.9.0`

**Comment by steven-sheehy** [comment]
My changes shouldn't affect the issue you're experiencing. `*` is expanded by bash not the s3 plugin. But `helm s3 push [<flags>] <chartPath> <repo>` doesn't take multiple charts paths so if bash happened to expand to multiple paths it wouldn't work.

**Comment by jgangemi** [comment]
weird b/c this used to work and `helm lint charts/*` earlier in the script has no issues. i'm sure i can change the script figure out the dirname.

thanks for responding and also doing this work, it saved me from having to figure out a hacky solution.

**Comment by rverma-nikiai** [comment]
@hypnoglow can we merge this pr

**Comment by jvosantos** [comment]
@hypnoglow is this going to be merged? it's been some months now

**Comment by cplee** [comment]
@steven-sheehy - do you have your fork hosted anywhere that I can move over to?  This PR looks abandoned...

**Comment by davidkarlsen** [comment]
Can this or #56 be merged? I use the plugin to publish the charts to a s3 bucket, which is then hosted by a cloudfront distro, hence I had to hack the index.yaml:
```yaml
     urls:
-    - s3://thebucketname/fspython-0.1.0.tgz
+    - fspython-0.1.0.tgz
```

**Comment by steven-sheehy** [comment]
I don't want to spend time fixing the conflicts (which are not trivial) unless there's some indication the PR has the possibility of being reviewed and accepted.

**Comment by davidkarlsen** [comment]
@hypnoglow care to comment?

**Comment by davidkarlsen** [comment]
Also - I think this can be simplified, by having the list in the url element only list the archive name - i.e. without the protocol&host prefix - that worked for me - and they are resolved relative to the repo given.
